### DISCUSSION TITLE
Will fastapi handle concurrent user?

### DISCUSSION THREAD

**Discussion by sindhubha** [fastapi/fastapi#10645] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@app.post("/demo/")
async def demo(cnx: Session = Depends(get_db)):
    query = (f'...')
    data = cnx.execute(query).fetchall()
```

### Description

where there are too many connections, application get's hanging .

### Operating System

Windows

### FastAPI Version

0.95.1

### Pydantic Version

1.10.7

### Python Version

3.9.12

**Comment by nymous** [comment, marked as answer]
Hello!

If your `cnx.execute(query).fetchall()` call is synchronous but is run in an async context (like your endpoint here which you marked as `async`), it will block the event loop and the entire application.
If your endpoint uses sync code, you should _not_ declare it as `async` and let FastAPI run it in a threadpool.
You could also use async versions of your dependencies. Your code looks like SQLAlchemy, which supports async now.

See https://fastapi.tiangolo.com/async/ for more information:
> If you are using a third party library that communicates with something (a database, an API, the file system, etc.) and doesn't have support for using `await`, (this is currently the case for most database libraries), then declare your path operation functions as normally, with just `def`

**Reply by sindhubha** [reply]
ok..

engine = create_engine("mysql+pymysql://root@localhost:3306/mydatabase",poolclass=NullPool)
SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)

def get_db():
   
    try:    
        db = SessionLocal()
        yield db
    finally:
        db.close() 

this is the connection code..
is mysql support async?
could you send me one sample api for better understand

**Reply by sriram-kondakindi** [reply]
```
import urllib.parse

from sqlalchemy.orm import sessionmaker, declarative_base
from sqlalchemy.ext.asyncio import create_async_engine, AsyncSession

from common.logger import logger
from common.exceptions import MyErrors, UnicornException
from config import settings

class DatabaseHelper:
    def __init__(self):
        try:
            self.engine = create_async_engine(f'mysql+aiomysql://{settings.sql_database_user}:'
                                              f'{urllib.parse.quote_plus(settings.sql_database_pass)}@'
                                              f'{settings.sql_database_host}:{settings.sql_database_port}/'
                                              f'{settings.sql_database_name}',
                                              pool_size=settings.sql_connection_pool_size,
                                              max_overflow=settings.sql_connection_pool_overflow,
                                              pool_recycle=settings.sql_pool_recycle_limit,
                                              pool_pre_ping=settings.sql_pool_pre_ping)
            self.MyBase = declarative_base()
            self.AsyncSession = sessionmaker(bind=self.engine, autocommit=settings.sql_autocommit,
                                             autoflush=settings.sql_autoflush,
                                             expire_on_commit=False, class_=AsyncSession)
        except Exception as e:
            logger.log(level=logger.error, msg=e)
            raise UnicornException(status_code=MyErrors.DATABASE_CONNECTION_FAILED[0],
                                   name=MyErrors.DATABASE_CONNECTION_FAILED[1] % (settings.sql_database_host,
                                                                                  settings.sql_database_user))


database_helper = DatabaseHelper()
```

and for creating the connection you can use the context manager directly in the function:

```
async with database_helper.AsyncSession() as session:
        try:
            pass
        except Exception as e:
            pass
        else:
            await session.commit()
```
I am not exactly aware on how to use it with fast api Dependency injection, but that should also be doable
